tests/vm_watch_test.go

```diff
@@ -21,338 +24,337 @@ const (
-			fallthrough
-		case flen > 3:
-			stat.ip = fields[3]
-			fallthrough
-		case flen > 2:
-			stat.phase = fields[2]
-			fallthrough
-		case flen > 1:
-			stat.age = fields[1]
-			fallthrough
-		case flen > 0:
-			stat.name = fields[0]
-		}
+	return "", nil
+}
 
-		return stat
-	}
+// Reads from rc until a newline character is found
+func readLineWithTimeout(rc io.ReadCloser, timeout time.Duration) (string, error) {
+	lineChan := make(chan string)
+	errChan := make(chan error)
+	defer close(errChan)
 
-	// Fail the test if stderr has something to read
-	failOnError := func(rc io.ReadCloser) {
-		defer GinkgoRecover()
+	go func() {
+		var line strings.Builder
+		buf := make([]byte, 1)
+		defer close(lineChan)
 
-		buf := make([]byte, bufferSize)
+		for {
+			n, err := rc.Read(buf)
 
-		n, err := rc.Read(buf)
+			if err != nil && err != io.EOF && !strings.Contains(err.Error(), "file already closed") {
+				errChan <- err
+				return
+			}
 
-		if err != nil && n > 0 {
-			rc.Close()
-			Fail(string(buf[:n]))
+			if n > 0 {
+				if buf[0] != '\n' {
+					line.WriteByte(buf[0])
+				} else {
+					break
+				}
+			}
 		}
-	}
-
-	// Reads from stdin until a newline character is found
-	readLine := func(rc io.ReadCloser, timeout time.Duration) string {
-		lineChan := make(chan string)
-
-		go func() {
-			defer GinkgoRecover()
-
-			var line strings.Builder
-			buf := make([]byte, 1)
-			defer close(lineChan)
 
-			for {
-				n, err := rc.Read(buf)
+		lineChan <- line.String()
+	}()
 
-				if err != nil && err != io.EOF && !strings.Contains(err.Error(), "file already closed") {
-					Fail(err.Error())
-					return
-				}
-
-				if n > 0 {
-					if buf[0] != '\n' {
-						line.WriteByte(buf[0])
-					} else {
-						break
-					}
-				}
-			}
+	select {
+	case line := <-lineChan:
+		return line, nil
+	case err := <-errChan:
+		return "", err
+	case <-time.After(timeout):
+		return "", fmt.Errorf("timeout reached on read operation")
+	}
+}
 
-			lineChan <- line.String()
-		}()
+// Reads VM/VMI status from rc and returns a new status.
+// If oldStatus is non-nil, the function will read status lines until
+// newStatus.running != oldStatus.running or newStatus.phase != oldStatus.phase
+// in order to skip duplicated status lines
+func readNewStatus(rc io.ReadCloser, oldStatus []string, timeout time.Duration) ([]string, error) {
+	statusLine, err := readLineWithTimeout(rc, timeout)
 
-		select {
-		case line := <-lineChan:
-			return line
-		case <-time.After(timeout):
-			err := rc.Close()
-			Expect(err).ToNot(HaveOccurred(), "stdout should have been closed properly")
+	if err != nil {
+		return nil, err
+	}
 
-			Fail("Timeout reached on read operation")
+	newStatus := strings.Fields(statusLine)
 
-			return ""
+	// Skip status line with similar running state for VM or phase state for VMI
+	// newStatus[2] and oldStatus[2] point to the VM running state or VMI phase
+	if oldStatus != nil {
+		if len(oldStatus) == 2 && len(oldStatus) == len(newStatus) {
```

Comment:
`len(oldStatus) == len(newStatus)` can be refactored to the parent `if`

Reply:
You are right, thanks!